### PR TITLE
Android phase 2 (second try)

### PR DESCRIPTION
**Pull request by mhsmith**
Before merging, please review my changes to the back-end code in `lib`. I think they should all be self-explanatory except for `Network.set_parameters`. I've split that into two parts: one to save to the config, and the other to apply the config. That way each part can be called separately by the UI.

Instructions for building the app are in `android/README.md`. Jonald has a copy of the release key.

### CODE REVIEW

**Comment by cculianu** [comment]
Ok, thanks man.  Sorry for the merge confusion.

@fyookball Give me a few hours at least to look this over. Thanks.

File: lib/exchange_rate.py [hunk 1]
```diff
@@ -15,6 +15,10 @@
 from .util import PrintError, ThreadJob
 
 
+DEFAULT_ENABLED = False
+DEFAULT_CURRENCY = "EUR"
+DEFAULT_EXCHANGE = "Kraken"
```
Comment (cculianu) [inline lib/exchange_rate.py]: Yes. Good. :)

File: lib/exchange_rate.py [hunk 2]
```diff
@@ -363,17 +367,18 @@ def set_fiat_address_config(self, b):
...
-        return self.config.get('use_exchange', 'Kraken')
+        return self.config.get('use_exchange', DEFAULT_EXCHANGE)
 
     def show_history(self):
         return self.is_enabled() and self.get_history_config() and self.ccy in self.exchange.history_ccys()
 
     def set_currency(self, ccy):
         self.ccy = ccy
-        self.config.set_key('currency', ccy, True)
+        if self.get_currency() != ccy:
```
Comment (cculianu) [inline lib/exchange_rate.py]: Nice little performance tune-up I guess?

File: lib/exchange_rate.py [hunk 3]
```diff
@@ -403,8 +408,12 @@ def exchange_rate(self):
             return Decimal(rate)
 
     def format_amount_and_units(self, btc_balance):
+        amount_str = self.format_amount(btc_balance)
+        return '' if not amount_str else "%s %s" % (amount_str, self.ccy)
+
+    def format_amount(self, btc_balance):
```
Comment (cculianu) [inline lib/exchange_rate.py]: Fair enough.

File: lib/interface.py [hunk 4]
```diff
@@ -274,6 +274,12 @@ def __init__(self, server, socket):
         
         self.mode = None
         
+    def __repr__(self):
+        return "<{}.{} {}>".format(__name__, type(self).__name__, self.format_address())
+
+    def format_address(self):
+        return "{}:{}".format(self.host, self.port)
```
Comment (cculianu) [inline lib/interface.py]: 👍

File: lib/simple_config.py [hunk 5]
```diff
@@ -157,12 +157,12 @@ def convert_version_2(self):
...
 
         try:
-            # migrate server string FROM host:port:proto TO host:port
+            # change server string FROM host:port:proto TO host:port:s
             server_str = self.user_config.get('server')
             host, port, protocol = str(server_str).rsplit(':', 2)
             assert protocol in ('s', 't')
             int(port)  # Throw if cannot be converted to int
-            server_str = str('{}:{}'.format(host, port))
+            server_str = str('{}:{}:s'.format(host, port))
```
Comment (cculianu) [inline lib/simple_config.py]: I'm a little confused about this one.  Why would "convert_version_2" need to be rewritten to always hard-code ssl ports? Huh? Why this change.  Convince me it won't break anything...

Also some servers *are* non-ssl. Not clear on why the hardcoded 's' there.

Please elaborate if you can. Thanks.

File: lib/transaction.py [hunk 6]
```diff
@@ -695,13 +695,14 @@ def get_fee(self):
...
-        return len(self.serialize(True)) / 2 if not self.is_complete() or self.raw is None else len(self.raw) / 2 # ASCII hex string
+        return (len(self.serialize(True)) // 2 if not self.is_complete() or self.raw is None
+                else len(self.raw) // 2)  # ASCII hex string
 
     @classmethod
     def estimated_input_size(self, txin):
         '''Return an estimated of serialized input size in bytes.'''
         script = self.input_script(txin, True)
-        return len(self.serialize_input(txin, script, True)) / 2
+        return len(self.serialize_input(txin, script, True)) // 2  # ASCII hex string
```
Comment (cculianu) [inline lib/transaction.py]: LOL.  Wonder why they initially were ok with the float there.  :)

File: lib/wallet.py [hunk 7]
```diff
@@ -468,7 +468,7 @@ def get_tx_height(self, tx_hash):
                 return height, conf, timestamp
             else:
                 height = self.unverified_tx[tx_hash]
-                return height, 0, False
```
Comment (cculianu) [inline lib/wallet.py]: Weird that they had that in there.  Good fix.

File: lib/wallet.py [hunk 8]
```diff
@@ -1421,6 +1421,7 @@ def delete_address(value):
         requests = {addr.to_storage_string() : delete_address(value.copy())
                     for addr, value in self.receive_requests.items()}
         self.storage.put('payment_requests', requests)
+        self.storage.write()
```
Comment (cculianu) [inline lib/wallet.py]: I seem to recall getting bugs related to this on iOS.  I actually had to explicitly call `wallet.storage.write()` on iOS to deal with this issue.  At the time I wanted iOS to be as "non-intrusive" as possible so I didn't bother to issue a PR to fix this. I figured the devs that wrote this knew what they were doing.

Good on you for adding this. it always bugged me. :)

File: lib/wallet.py [hunk 9]
```diff
@@ -1538,8 +1540,9 @@ def can_change_password(self):
     def update_password(self, old_pw, new_pw, encrypt=False):
         if old_pw is None and self.has_password():
             raise InvalidPassword()
-        self.keystore.update_password(old_pw, new_pw)
-        self.save_keystore()
+        if self.keystore is not None:
```
Comment (cculianu) [inline lib/wallet.py]: Is it ever None?

File: lib/simple_config.py (hunk 5 above)
Comment (mhsmith) [reply lib/simple_config.py]: The string is actually required to have 3 fields, and `network.deserialize_server` will crash if it only has 2.  So the previous code was invalid.

Commit cba8052e7501803ebc49d458568dd96c14c7960b (cherry-picked from Electrum commit f77f029d728456e2a080abd7cf688afc270a20bc) was supposed to fix this, but somehow failed to include both files.

I don't know why those committers wanted to hard-code the 's', but it was clearly their intention.

File: lib/transaction.py (hunk 6 above)
Comment (cculianu) [reply lib/transaction.py]: Ah yeah, true.

File: lib/exchange_rate.py (hunk 2 above)
Comment (cculianu) [reply lib/exchange_rate.py]: Ah ha -- great approach to do that.  That's pretty cool. Makes even more sense then. 👍

File: lib/simple_config.py (hunk 5 above)
Comment (cculianu) [reply lib/simple_config.py]: Right but the thing that gets me is:

- they parse the protocol; 's' or 't' 2 lines up.  
- 's' and 't' are still supported by extant code. 
- They basically throw away information and force 't' servers now to all be 's'.. why?

I'm going to ask ghost43 about this.  You know.. he could have just made a mistake. :)

Comment (mhsmith) [reply lib/simple_config.py]: I don't understand it either, and yet the code, the comment, and the commit message all agree. Let's see what he says.

**Comment by cculianu** [comment]
Ok well ghost43 didn't get back to me.  Your changes to simple_config.py are at least less harmful than what was there already.  It doesn't matter anyway -- it's just for upgrades from v1.

Probably v1 only supported 's' servers?

No idea.  Merging.

**Comment by mhsmith** [comment]
Response from ghost43:

> it's more meaningful if you also consider some related commits. IIRC that's when the "use SSL" checkbox was removed from the network dialog. we wanted to "soft" enforce that everyone uses SSL to connect to the server, but still allow advanced users to  use plaintext. so the checkbox was removed, and you could no longer "change the protocol" in the GUI. and advanced users can use CLI options or edit the config to use TCP. however, what if some users had unchecked the "Use SSL" checkbox in older versions, that got saved in the config, and now they would be "stuck" not using SSL... so a "config upgrade" was implemented that changed the protocol saved in the config to SSL. chances are, if you had a TCP server configured and then upgraded, the same host+port combo with SSL was not usable, however, you would have the option to connect to another server (or it would happen automatically with auto-connect), and the larger issue of getting stuck on TCP was averted

**Comment by cculianu** [comment]
Ah yeah I saw that.  Good to have it here for the record. Makes sense! 👍
